### DISCUSSION TITLE
Compatible Axi4 formal VIP or partial one?

### DISCUSSION THREAD

**Discussion by Readon** [SpinalHDL/SpinalHDL#732] [General]
I am thinking about the implementation of Axi4 formal VIP.
As the spec said, write data channel can be independent to write address channel, so that write data operation can be complete before write address. 
However, that would make the verification helper logic too complicated.
As I understand, it would involve FIFOs for two channels, which could lead "prove" hard.

If the assumption that write data must act after or at the same time with address channel, then things are getting simplified.
In spec, it also said if the interconnect is involved, then write data and address operations should be realigned.

**Comment by Dolu1990** [comment]
> In spec, it also said if the interconnect is involved, then write data and address operations should be realigned.

Should ? should, like : has the obligation ? 

I mean, i understand that the interconnect has at some point to sync up things to know what to do, but is it required by the spec ?

**Reply by Readon** [reply]
There is a note for that in A3.3 of Axi4 spec.
Here is some understanding(I don't know if I can put the text directly here.): 
     When the interconnect module needs to determine the destination address space, it must realign the address and write data to ensure that the write data channel only direct to the right destination.

it's a "must".

**Reply by Dolu1990** [reply]
Ahhh, yes, so, for logical purposes. but we can't assume anything on the axi interface going out of the interconnect. The reason is that even if it internaly realign things for routing reasons, it may contains some pipelineing stages after the realign, which would break the assumption.

Note the alignement is only necessary in the decoder, not the  arbiter side.

**Reply by Readon** [reply]
So it becomes complicated when data's arrival is independent to address and length's.
Because it can be asserted before or after address with length, especially when data arrived ahead of length.
Is there any good idea about this?

**Reply by Dolu1990** [reply]
what could be done, maybe, is to add a buffer on the data channel (StreamFifo but with a assumption that it never goes full) in the verification agent.

That way, you can reorder things as you wish :)
Would that be good ?

**Comment by Readon** [comment]
I have checked what ARM VIP do and find that they are using a CAM to save the w data and aw data. The elements of those CAM can be removed in any position.&nbsp;&nbsp;
I think it is a kind of History with pop function as I just committed in #721.



---Original---
From: ***@***.***&gt;
Date: Mon, May 16, 2022 18:12 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [SpinalHDL/SpinalHDL] Compatible Axi4 formal VIP or partial one? (Discussion #732)




 
what could be done, maybe, is to add a buffer on the data channel (StreamFifo but with a assumption that it never goes full) in the verification agent.
 
That way, you can reorder things as you wish :)
 Would that be good ?
